File: frontend/src/api/projects.ts
import api from './client';

// ==================== Types ====================

export interface Task {
  id: string;
  project_id: string;
  parent_id: string | null;
  title: string;
  start_date: string;
  end_date: string;
  is_done: boolean;
  sort_order: number;
  created_at: string;
  children: Task[];
}

export interface Project {
  id: string;
  name: string;
  sort_order: number;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
  tasks?: Task[];
}

// ==================== Projects ====================

export const getProjects = async (): Promise<Project[]> => {
  const res = await api.get('/projects/');
  return res.data;
};

export const getArchivedProjects = async (): Promise<Project[]> => {
  const res = await api.get('/projects/archived');
  return res.data;
};

export const createProject = async (name: string): Promise<Project> => {
  const res = await api.post('/projects/', { name });
  return res.data;
};

export const updateProject = async (id: string, data: Partial<Project>): Promise<Project> => {
  const res = await api.put(`/projects/${id}`, data);
  return res.data;
};

export const deleteProject = async (id: string): Promise<void> => {
  await api.delete(`/projects/${id}`);
};

export const archiveProject = async (id: string): Promise<Project> => {
  const res = await api.put(`/projects/${id}/archive`);
  return res.data;
};

export const unarchiveProject = async (id: string): Promise<Project> => {
  const res = await api.put(`/projects/${id}/unarchive`);
  return res.data;
};

export const reorderProjects = async (ids: string[]): Promise<void> => {
  await api.put('/projects/reorder', { ids });
};

// ==================== Tasks ====================

export const getTasks = async (projectId: string): Promise<Task[]> => {
  const res = await api.get(`/projects/${projectId}/tasks`);
  return res.data;
};

export const createTask = async (projectId: string, data: {
  title: string;
  start_date: string;
  end_date: string;
  parent_id?: string | null;
  sort_order?: number;
}): Promise<Task> => {
  const res = await api.post(`/tasks/${projectId}`, data);
  return res.data;
};

export const updateTask = async (taskId: string, data: Partial<Task>): Promise<Task> => {
  const res = await api.put(`/tasks/${taskId}`, data);
  return res.data;
};

export const deleteTask = async (taskId: string): Promise<void> => {
  await api.delete(`/tasks/${taskId}`);
};

export const toggleTask = async (taskId: string): Promise<Task> => {
  const res = await api.put(`/tasks/${taskId}/toggle`);
  return res.data;
};

export const reorderTasks = async (items: { id: string; sort_order: number; parent_id?: string | null }[]): Promise<void> => {
  await api.put('/tasks/reorder', { items });
};

File: backend/app/routers/projects.py
import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from ..database import get_db
from .. import crud, schemas

router = APIRouter()


@router.get("/")
def list_projects(db: Session = Depends(get_db)):
    projects = crud.get_projects(db)
    return projects


@router.get("/archived")
def list_archived_projects(db: Session = Depends(get_db)):
    projects = crud.get_archived_projects(db)
    return projects


@router.post("/")
def create_project(data: schemas.ProjectCreate, db: Session = Depends(get_db)):
    return crud.create_project(db, data.name)


@router.put("/reorder")
def reorder_projects(data: schemas.ProjectReorder, db: Session = Depends(get_db)):
    crud.reorder_projects(db, data.ids)
    return {"ok": True}


@router.put("/{project_id}")
def update_project(project_id: str, data: schemas.ProjectUpdate, db: Session = Depends(get_db)):
    result = crud.update_project(db, uuid.UUID(project_id), data)
    if not result:
        raise HTTPException(status_code=404, detail="Project not found")
    return result


@router.delete("/{project_id}")
def delete_project(project_id: str, db: Session = Depends(get_db)):
    if not crud.delete_project(db, uuid.UUID(project_id)):
        raise HTTPException(status_code=404, detail="Project not found")
    return {"ok": True}


@router.put("/{project_id}/archive")
def archive_project(project_id: str, db: Session = Depends(get_db)):
    result = crud.archive_project(db, uuid.UUID(project_id), True)
    if not result:
        raise HTTPException(status_code=404, detail="Project not found")
    return result


@router.put("/{project_id}/unarchive")
def unarchive_project(project_id: str, db: Session = Depends(get_db)):
    result = crud.archive_project(db, uuid.UUID(project_id), False)
    if not result:
        raise HTTPException(status_code=404, detail="Project not found")
    return result


@router.get("/{project_id}/tasks")
def list_tasks(project_id: str, db: Session = Depends(get_db)):
    return crud.get_tasks(db, uuid.UUID(project_id))
